# release notes

## current release (v6)
- Data release data: 2023-09-08
- OpenPedCan data release date: 2023-04-30 (v12)
- status: available

Additional files:
- `morpholno.merged.rmats.tsv` : merged matrix of rMATs splice events, comparing morphilino treated vs untreated KNS42 cells
- `ctrl_vs_morpho.rsem.genes.results.tsv`: count data matrix of morphilino treated and untreated KNS42 cells
- `CLK1_CRISPR_depmap_score.csv` : DepMap portal database file for dependency
- `OmicsDefaultModelProfiles.csv` : Profile mappings of cell lines and omics data

```
v6
├── cnv-consensus.seg.gz
├── consensus_wgs_plus_cnvkit_wxs.tsv.gz
├── fusion-putative-oncogenic.tsv
├── gene-counts-rsem-expected_count-collapsed.rds
├── gene-expression-rsem-tpm-collapsed.rds
├── histologies.tsv
├── independent-specimens.rnaseqpanel.primary-plus.tsv
├── independent-specimens.rnaseqpanel.primary.tsv
├── independent-specimens.wgswxspanel.primary-plus.prefer.wgs.tsv
├── independent-specimens.wgswxspanel.primary-plus.prefer.wxs.tsv
├── independent-specimens.wgswxspanel.primary.prefer.wgs.tsv
├── independent-specimens.wgswxspanel.primary.prefer.wxs.tsv
├── md5sum.txt
├── rMATS_merged.comparison.tsv.gz
├── release-notes.md
├── rna-isoform-expression-rsem-tpm.rds
├── snv-consensus-plus-hotspots.maf.tsv.gz
├── snv-mutation-tmb-coding.tsv
├── splice-events-rmats.tsv.gz
├── morpholno.merged.rmats.tsv
├── ctrl_vs_morpho.rsem.genes.results.tsv
├── CLK1_CRISPR_depmap_score.csv
└── OmicsDefaultModelProfiles.csv
```

## previous release (v5)
- Data release data: 2023-05-26
- OpenPedCan data release date: 2023-04-30 (v12)
- status: available

Additional files:
- `splice-events-rmats.tsv.gz`: merged matrix of rMATs splice events, run as single sample
- `rMATS_merged.comparison.tsv.gz`: Midline HGG sample splice events compared to non-diseased brainstem, run using rMATs
- `release-notes.md`: specific release notes to this repository

```
v5
├── cnv-consensus.seg.gz
├── consensus_wgs_plus_cnvkit_wxs.tsv.gz
├── fusion-putative-oncogenic.tsv
├── gene-counts-rsem-expected_count-collapsed.rds
├── gene-expression-rsem-tpm-collapsed.rds
├── histologies.tsv
├── independent-specimens.rnaseqpanel.primary-plus.tsv
├── independent-specimens.rnaseqpanel.primary.tsv
├── independent-specimens.wgswxspanel.primary-plus.prefer.wgs.tsv
├── independent-specimens.wgswxspanel.primary-plus.prefer.wxs.tsv
├── independent-specimens.wgswxspanel.primary.prefer.wgs.tsv
├── independent-specimens.wgswxspanel.primary.prefer.wxs.tsv
├── md5sum.txt
├── rMATS_merged.comparison.tsv.gz
├── release-notes.md
├── rna-isoform-expression-rsem-tpm.rds
├── snv-consensus-plus-hotspots.maf.tsv.gz
├── snv-mutation-tmb-coding.tsv
└── splice-events-rmats.tsv.gz
```

## previous release (v4)
- Data release data: 2023-04-30
- OpenPedCan data release date: 2023-04-30 (v12)
- status: available

Additional files:
- `splice-events-rmats.tsv.gz`: merged matrix of rMATs splice events, run as single sample

```
v4
├── cnv-consensus.seg.gz
├── consensus_wgs_plus_cnvkit_wxs.tsv.gz
├── fusion-putative-oncogenic.tsv
├── gene-counts-rsem-expected_count-collapsed.rds
├── gene-expression-rsem-tpm-collapsed.rds
├── histologies.tsv
├── independent-specimens.rnaseqpanel.primary-plus.tsv
├── independent-specimens.rnaseqpanel.primary.tsv
├── independent-specimens.wgswxspanel.primary-plus.prefer.wgs.tsv
├── independent-specimens.wgswxspanel.primary-plus.prefer.wxs.tsv
├── independent-specimens.wgswxspanel.primary.prefer.wgs.tsv
├── independent-specimens.wgswxspanel.primary.prefer.wxs.tsv
├── md5sum.txt
├── rna-isoform-expression-rsem-tpm.rds
├── snv-consensus-plus-hotspots.maf.tsv.gz
├── snv-mutation-tmb-coding.tsv
└── splice-events-rmats.tsv.gz
```
